src/clients/c++/inference_profiler.cc

```diff
@@ -0,0 +1,452 @@
+    : verbose_(verbose), profile_(profile),
+      measurement_window_ms_(measurement_window_ms),
+      max_measurement_count_(max_measurement_count),
+      extra_percentile_(extra_percentile), percentile_(percentile),
+      on_sequence_model_(on_sequence_model), model_name_(model_name),
+      model_version_(model_version), profile_ctx_(std::move(profile_ctx)),
+      status_ctx_(std::move(status_ctx)), manager_(std::move(manager))
+{
+  load_parameters_.stable_offset = stable_offset;
+}
+
+nic::Error
+InferenceProfiler::Profile(
+    const size_t concurrent_request_count, PerfStatus& status_summary)
+{
+  status_summary.concurrency = concurrent_request_count;
+
+  RETURN_IF_ERROR(manager_->ChangeConcurrencyLevel(concurrent_request_count));
+
+  // Start measurement
+  bool is_complete = true;
+  load_parameters_.infer_per_sec.clear();
+  load_parameters_.latencies.clear();
+  load_parameters_.avg_ips = 0;
+  load_parameters_.avg_latency = 0;
+
+  do {
+    RETURN_IF_ERROR(manager_->CheckHealth());
+
+    RETURN_IF_ERROR(Measure(status_summary));
+
+    load_parameters_.infer_per_sec.push_back(
+        status_summary.client_infer_per_sec);
+    load_parameters_.latencies.push_back(status_summary.stabilizing_latency_ns);
+    load_parameters_.avg_ips += (double)load_parameters_.infer_per_sec.back() /
+                                load_parameters_.stability_window;
+    load_parameters_.avg_latency +=
+        load_parameters_.latencies.back() / load_parameters_.stability_window;
+
+    if (verbose_) {
+      std::cout << "  Pass [" << load_parameters_.infer_per_sec.size()
+                << "] throughput: " << load_parameters_.infer_per_sec.back()
+                << " infer/sec. ";
+      if (extra_percentile_) {
+        std::cout << "p" << percentile_ << " latency: "
+                  << (status_summary.client_percentile_latency_ns
+                          .find(percentile_)
+                          ->second /
+                      1000)
+                  << " usec" << std::endl;
+      } else {
+        std::cout << "Avg latency: "
+                  << (status_summary.client_avg_latency_ns / 1000)
+                  << " usec (std " << status_summary.std_us << " usec)"
+                  << std::endl;
+      }
+    }
+
+    if (load_parameters_.infer_per_sec.size() >=
+        load_parameters_.stability_window) {
+      size_t idx = load_parameters_.infer_per_sec.size() -
+                   load_parameters_.stability_window;
+      if (load_parameters_.infer_per_sec.size() >
+          load_parameters_.stability_window) {
+        load_parameters_.avg_ips -=
+            (double)load_parameters_.infer_per_sec[idx - 1] /
+            load_parameters_.stability_window;
+        load_parameters_.avg_latency -= load_parameters_.latencies[idx - 1] /
+                                        load_parameters_.stability_window;
+      }
+      is_complete = manager_->IsLoadingComplete(load_parameters_);
```

Comment:
So now the profiler summarizes the result and then let the load generator to decide if the result is stable?

Reply:
Yes, For some types of loads, we might not be able to arrive at a stable average ips and latencies. Hence, profiler will send its recorded data to load manager and let it decide whether the load is stable. The logic behind coming up with a stopping criteria hence should be kept with each load manager.